// @forbes/fbs-assets-loader/dist/main.js
var loadedMap = new Map();
/**
 * Loads given script and calls callback onload
 */
export function loadScript(src, cb) {
    if (!src.includes('.js')) {
        throw new Error('You can only load JS files with this function.');
    }
    var script = loadedMap.get(src); // Grab an exsiting subject if available
    // If none existing, make a new one
    if (!script) {
        script = document.createElement('script');
        loadedMap.set(src, script);
        script.dataset.loaded = 'false';
        Object.assign(script, {
            async: true,
            defer: true,
            src: src
        });
        script.addEventListener('load', function() {
            return script.dataset.loaded = 'true';
        });
        script.addEventListener('load', cb);
        document.head.appendChild(script);
    } else if (script.dataset.loaded !== 'false') {
        cb();
    } else {
        script.addEventListener('load', cb);
    }
    return script;
}
/**
 * Loads given stylesheet and calls callback onload
 */
export function loadStyle(src, cb) {
    if (!src.includes('.css')) {
        throw new Error('You can only load CSS files with this function.');
    }
    var style = loadedMap.get(src); // Grab an exsiting style if available
    // If none existing, make a new one
    if (!style) {
        style = document.createElement('link');
        loadedMap.set(src, style);
        style.dataset.loaded = 'false';
        Object.assign(style, {
            href: src,
            rel: 'stylesheet',
            type: 'text/css'
        });
        style.addEventListener('load', function() {
            return style.dataset.loaded = 'true';
        });
        style.addEventListener('load', cb);
        document.head.appendChild(style);
    } else if (style.dataset.loaded !== 'false') {
        cb();
    } else {
        style.addEventListener('load', cb);
    }
    return style;
}


//////////////////
// WEBPACK FOOTER
// ./node_modules/@forbes/fbs-assets-loader/dist/main.js
// module id = 35
// module chunks = 16
