// @ux/notifs-v1/src/badge.js
import React from 'react';
import PropTypes from 'prop-types';
import { NamespaceConsumer, NamespaceShape, withNamespace } from '@ux/namespace-component';
import shapes from './shapes';

class NotificationBadge extends NamespaceConsumer {
  render() {
    const { count, severity, dismissed, tooltip, messages = {} } = this.props;
    const bellClasses = this.classNames(
      this.namespace('btn', 'dropdown-toggle'),
      this.namespace(this.classNames({
        'notifications-mini-bell': count,
        'notifications-bell': !count
      }))
    );
    const messagesLabel = tooltip || messages.headerText;

    if (count > 0) {
      const badgeClasses = this.classNames(
        this.namespace('notifications-badge', 'font-primary-bold'),
        this.namespace(this.classNames({ dismissed, [severity]: !dismissed })) // backwards compatibility with eldorado
      );

      return (
        <button aria-label={ messagesLabel } className={ bellClasses } title={ tooltip } type='button'>
          <div className={ badgeClasses }>
            { count }
          </div>
        </button>
      );
    }

    return (
      <button aria-label={ messagesLabel } className={ bellClasses } title={ tooltip } type='button' />
    );
  }
}

NotificationBadge.propTypes = {
  ...NamespaceShape,
  count: PropTypes.number,
  severity: PropTypes.string,
  dismissed: PropTypes.bool,
  tooltip: PropTypes.string,
  messages: shapes.messages
};

NotificationBadge.defaultProps = {
  count: 0,
  severity: 'none'
};

export default withNamespace(NotificationBadge);
