// @guardian/consent-management-platform/lib/cookies.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
var Cookies = require("js-cookie");
var config_1 = require("./config");
var getShortDomain = function() {
    var domain = document.domain || '';
    return domain === 'localhost' ?
        domain :
        [''].concat(domain.split('.').slice(-2)).join('.');
};
var getDomainAttribute = function() {
    var shortDomain = getShortDomain();
    return shortDomain === 'localhost' ? '' : shortDomain;
};
var addCookie = function(name, value) {
    var options = {
        domain: getDomainAttribute(),
        expires: config_1.COOKIE_MAX_AGE,
    };
    Cookies.set(name, value, options);
};
var readGuCookie = function() {
    var cookie = Cookies.getJSON(config_1.GU_COOKIE_NAME);
    if (cookie) {
        if (cookie.version === 1) {
            return cookie.state || null;
        }
    }
    return null;
};
exports.readGuCookie = readGuCookie;
var readIabCookie = function() {
    var cookie = Cookies.get(config_1.IAB_COOKIE_NAME);
    return cookie || null;
};
exports.readIabCookie = readIabCookie;
var writeGuCookie = function(guState) {
    return addCookie(config_1.GU_COOKIE_NAME, {
        version: config_1.GU_COOKIE_VERSION,
        state: guState
    });
};
exports.writeGuCookie = writeGuCookie;
var writeIabCookie = function(iabString) {
    return addCookie(config_1.IAB_COOKIE_NAME, iabString);
};
exports.writeIabCookie = writeIabCookie;
