// @ux/sales-footer/index.js
import React from 'react';
import PropTypes from 'prop-types';
import Button from '@ux/button';
import Modal from '@ux/modal';
import Tooltip from '@ux/tooltip';
import CurrencySelector from '@ux/currency-selector';
import MarketSelector from '@ux/market-selector';
import Logo from '@ux/logo';
import EmailForm from './email-form';
import { NamespaceProvider, NamespaceConsumer, NamespaceShape, withNamespace } from '@ux/namespace-component';
import { IntlProvider, FormattedMessage, FormattedHTMLMessage } from 'react-intl';
import { breakpoints } from './enums';
import assign from 'object-assign';
import Link from '@ux/link';
import golfToEid from 'golf-to-eid';
import WeChatIcon from './wechat-icon.js';


/**
 * Return a namespace if the environment variable is set, 'polyfilled' by WebPack.
 *
 * @returns {Object} Configured namespace.
 * @api private
 */
function namespace() {
  return process.env.NAMESPACE; // eslint-disable-line no-process-env
}

const vatInclusionCodes = {
  ALL: 'Sales:Footer:VAT:Disclaimer',
  NONE: 'Sales:Footer:VAT:DisclaimerNone',
  TAXES: 'Sales:Footer:VAT:DisclaimerTaxes',
  ICANN: 'Sales:Footer:VAT:DisclaimerICANN'
};

/**
 * Render and expose the Product navigation as a Component.
 *
 * @param {Object} props Properties
 * @returns {ReactElement} Namespaced Product Nav.
 * @implements {React.Component}
 * @api public
 */
class SalesFooterComponent extends NamespaceConsumer {
  constructor(props) {
    super(...arguments);

    this.state = {
      ...this.state,
      ...this.nextStateFromProps(props),
      goDaddyStoreModal: false,
      selected: null
    };
  }

  /**
   * Prefix for generated EIDs
   *
   * @type {String}
   * @readonly
   * @private
   */
  get eidPrefix() {
    return this.props.seechange ? 'uxp.hyd.sales_footer_seechange' : 'uxp.hyd.sales_footer';
  }

  get footerLinkEid() {
    return {
      prefix: this.eidPrefix,
      postfix: 'link.click'
    };
  }

  /**
   * construct an li wrapped a tag from a manifest-render-service navigation item
   *
   * @param {Object} items structure containing link data
   * @param {Object} options options structure
   * @returns {Array} a configured anchor tag
   * @private
   */
  linkListBuilder(items = [], options = {}) {
    const className = [this.classNames(options.className)];
    const active = options.active || false;

    return items.map((item, i) => {
      const local = className.slice();
      if (active && ~active.indexOf(item.id)) {
        local.push(this.namespace('active'));
      }

      let onClick;
      if (item.caption === 'Sales:Footer:AboutGD:GoDaddyStore' || (item.id && ~item.id.indexOf('GoDaddyStore'))) {
        onClick = this.showGoDaddyStoreModal.bind(this);
      }

      // `title` refers to the `title` attribute on the anchor in Link, so we need to strip it here and pass it as a
      // child if present
      const itemWithoutTitle = {
        ...item,
        title: void 0
      };
      const linkChildren = item.title ? <span>{ item.title }</span> : void 0;
      const captionOrId = item.caption || item.id;
      return (
        <p key={ captionOrId + i } className={ local.filter(Boolean).join(' ') }>
          <Link
            { ...itemWithoutTitle }
            { ...options }
            href={ item.href || '#' }
            onClick={ onClick }
            data-eid={ golfToEid(captionOrId, this.footerLinkEid) }>
            { linkChildren }
          </Link>
        </p>
      );
    });
  }

  /**
   * Updates the state associated with this instance to display a modal
   * with content related to the GoDaddy Store
   * @param   {Event} evt React Synthetic event for Click.
   * @returns {Boolean} Value indicating to stop propagation.
   */
  showGoDaddyStoreModal(evt) {
    evt.preventDefault();
    this.setState({ goDaddyStoreModal: true });
    return false;
  }

  /**
   * Returns a React Element to display the GoDaddy Store model if necessary
   * @returns {React.ReactNode} The UXCore2 modal or null (if appropriate).
   */
  renderGoDaddyStoreModal() {
    if (!this.state.goDaddyStoreModal) {
      return null;
    }

    // 'Shared:Common:OK' isn't localized in eldorado.
    // Remark: do we need to make a new string for "OK"?
    const title = <FormattedMessage id='PC:Modal:GodaddyStore:HeaderTitle' />;
    const footer = (
      <Button
        design='primary'
        onClick={ () => {
          window.open(this.props.urls.store.href);
          this.setState({ goDaddyStoreModal: false });
        } }
        data-eid='uxp.hyd.sales_footer.godaddy_store.modal.ok.link.click'>
        <FormattedMessage id='Shared:Common:OK' />
      </Button>
    );

    return (
      <Modal
        title={ title }
        onClose={ () => { this.setState({ goDaddyStoreModal: false }); } }
        footer={ footer }>
        <p>
          <FormattedMessage id='PC:Modal:GodaddyStore:HeaderText' />
        </p>
      </Modal>
    );
  }

  /**
   * @param {Object} currItem containing link data
   * Returns a React Element to display the WeChat Tooltip if necessary
   * @returns {React.ReactNode} The UXCore2 FormattedMessage with Tooltip.
   */
  renderWeChat(currItem) {
    const props = this.props;
    const market = props.market;
    const cdn = props.urls &&
      props.urls.cdn &&
      props.urls.cdn.href;

    const { imgPath } = currItem.options || {};

    const weChatQR = (altText) => <img
      className={ this.namespace('wechat-qr') }
      src={ `${cdn}${imgPath.replace('{market}', market.toLowerCase())}` }
      height='247px'
      width='218px'
      alt={ altText }/>;

    return (
      <FormattedMessage id={ 'Sales:Footer:SocialMedia:WeChatIcon' }>
        { text => (
          <Tooltip
            data-eid={ golfToEid(currItem.caption || currItem.id, {
              prefix: 'uxp.hyd.sales_footer',
              postfix: 'button.click'
            }) }
            className='footer-wechat'
            aria-label={ text }
            tabIndex='0'
            id='footer-wechat'
            text={ <FormattedMessage id={ 'Sales:Footer:SocialMedia:WeChatIcon' }>
              { altText => (<WeChatIcon className={ this.namespace('wechat-icon') } altText={ altText }/>)}
            </FormattedMessage> }
            message={ currItem.caption ? <IntlProvider locale={ props.market } messages={ props.messages }>
              <FormattedMessage id={ currItem.caption }>
                { weChatQR }
              </FormattedMessage>
            </IntlProvider> : weChatQR(currItem.title) }
            title={ currItem.caption ? <IntlProvider locale={ props.market } messages={ props.messages }>
              <FormattedMessage id={ currItem.caption } /></IntlProvider> : currItem.title }>
          </Tooltip>
        )}
      </FormattedMessage>
    );
  }


  /**
   * Renders links in the legal section
   *
   * @param {Object}   props Full set of props for this component.
   * @param {Object[]} links Array of objects that have a caption and href property
   * @param {String} [liClass] Optional classname to include on li elements
   * @api private
   * @returns {React.ReactNode} li for each item in links
   */
  buildLegalLinks(props, links, liClass) {
    return links.map((link, index) => {

      // plid=3153 is used when "we do not know" the plid in which case we remove the privacy policy link.
      if (props.privateLabelId === 3153 &&
        ~['Shared:LegalBar:LinkList:PrivacyPolicy',
          'Shared:LegalBar:LinkList:ResellerPrivacyPolicy'].indexOf(link.caption)) {
        return;
      }

      return (
        <li key={ index } className={ liClass }>
          <Link
            urlArgs={ props.urlArgs }
            href={ link.href }
            data-eid={ golfToEid(link.caption || link.id, this.footerLinkEid) }>
            { link.caption ? <FormattedMessage id={ link.caption } /> : link.title }
          </Link>
        </li>
      );
    });
  }

  selectFooterCol(selected) {
    // we should not be able to toggle in a screen larger than a phablet
    // expand/collapse is no longer an option with seechange
    if (this.props.seechange || !this.isMobile) {
      return;
    }

    // just close if we select something currently open
    if (selected === this.state.selected) {
      return this.setState({ selected: null });
    }

    this.setState({ selected });
  }

  /**
   * Render footer links in sections
   *
   * @param   {Object} props Full set of props for this component.
   * @returns {React.ReactNodeArray} A set of <Dropdown/> components.
   * @api private
   */
  buildFooterLinks(props) {
    const { seechange } = props;
    const upperLinks = props.footer.filter(item => {
      const captionOrId = item.caption || item.id;
      if (captionOrId !== 'Sales:Footer:MobileApps:Heading' && captionOrId !== 'Sales:Footer:SocialMedia:Heading') {
        return item;
      }
    });

    const navItemChildren = upperLinks.map(item => {
      return this.linkListBuilder(item.children, {
        className: this.namespace('footernav-item'),
        active: props.active,
        urlArgs: props.urlArgs
      });
    });

    return upperLinks.map((curr, index) => {
      const opened = !this.isMobile || props.seechange || this.state.selected === index;
      const columnBreak = ((index + 1) % 2 === 0 && seechange) ?
        <div className={ this.namespace('hidden-sm-down', 'hidden-lg-up', 'clearfix') }></div> :
        null;
      return (
        <React.Fragment key={ index }>
          <div className={ this.namespace(
            !seechange ? 'col-md-2' : 'col-lg-2',
            seechange ? 'col-md-6' : '',
            'col-sm-12', 'col-12',
            'footer-links-column',
            opened ? 'opened' : 'closed') }>
            <button
              type='button'
              onClick={ () => this.selectFooterCol(index) }
              className={ this.namespace('column-heading', 'font-primary-bold') }>
              { curr.caption ? <FormattedMessage id={ curr.caption } /> : curr.title }
            </button>
            { navItemChildren[index] }
          </div>
          { columnBreak }
        </React.Fragment>
      );
    });
  }

  /**
 * Builds a set of links that have corresponding images
 * @param  {Object} items Object that contains an array of children with link data
 * @param  {String} [spanClass] Class to be added to the span.  Must be pre-namespaced by the caller.
 * @param  {Object} [urlArgs]   Args to append as a query to the link
 * @returns {React.ReactElement[]} </li> for each link with an </img>
 * @api private
 */
  buildLinksWithImg(items, spanClass, urlArgs) {
    const props = this.props;
    const cdn = props.urls &&
      props.urls.cdn &&
      props.urls.cdn.href;
    if (!cdn || !Array.isArray(items.children)) {
      return [];
    }

    return items.children.map((currItem, index) => {
      // className comes from `eldorado-data`
      const { className, imgPath } = currItem.options || {};

      if (!imgPath) return;
      const imgFromAlt = (altText) => <img src={ `${cdn}${imgPath}` } alt={ altText }/>;

      if (currItem.caption === 'Sales:Footer:SocialMedia:WeChatQRCode' ||
        currItem.id === 'Sales:Footer:SocialMedia:WeChatQRCode') {
        return (
          <li key={ index }>
            { this.renderWeChat(currItem) }
          </li>
        );
      }

      return (
        <li key={ index }>
          <Link urlArgs={ urlArgs }
            href={ currItem.href }
            data-eid={ golfToEid(currItem.caption || currItem.id, {
              prefix: 'uxp.hyd.sales_footer',
              postfix: 'link.click'
            }) }>
            <span className={ this.classNames(spanClass, this.namespace(className)) }>
              { currItem.caption ? <FormattedMessage id={ currItem.caption }>
                {imgFromAlt}
              </FormattedMessage> : imgFromAlt(currItem.title) }
            </span>
          </Link>
        </li>
      );
    });
  }

  /**
 * Renders the div for the app store images/links
 * @returns {React.ReactElement} Div containing app store badges
 * @api private
 */
  renderAppStoreSection() {
    const appStores = this.props.footer.filter(x => x.caption === 'Sales:Footer:MobileApps:Heading' ||
      (x.id && ~x.id.indexOf('MobileApps')))[0];
    if (!appStores) return;
    return (
      <div className={ this.namespace('app-store-section') }>
        <span className={ this.namespace('font-primary-bold') }>
          { appStores.caption ? <FormattedMessage id={ appStores.caption } /> : appStores.title }
        </span>
        <ul className={ this.namespace('list-inline') }>
          { this.buildLinksWithImg(appStores) }
        </ul>
      </div>
    );
  }

  /**
   * Returns a React Element representing the social media links for the specified `props`.
   *
   * @param   {Object} props Full set of props for this component.
   * @returns {React.ReactElement[]} <li/> for each item in props.footer that is a child of Sales:Footer:SocialMedia:Heading
   * @api private
   */
  buildSocialMediaLinks(props) {
    const socialMedia = props.footer.filter(x => x.caption === 'Sales:Footer:SocialMedia:Heading' ||
      (x.id && ~x.id.indexOf('SocialMedia')))[0];
    if (!socialMedia) {
      return [];
    }

    return this.buildLinksWithImg(socialMedia, this.namespace('footer-social-icon'), props.urlArgs);
  }

  /**
   * Renders the legal section.
   *
   * @param   {Object} props Full set of props for this component.
   * @returns {React.ReactNode} div containing disclaimer, legal links, and copyright
   * @api private
   */
  buildLegalSection(props) {
    const sections = {};
    const rows = props.legal.map((section, index) => { // eslint-disable-line max-statements
      let firstCaption = '';
      let firstChild = {};
      if (section.children) {
        firstChild = section.children[0];
        firstCaption = firstChild.caption || firstChild.id;
      }
      const sectionCaptionOrId = section.caption || section.id;

      // Legal disclaimer
      if (~firstCaption.indexOf('Disclaimer')) {
        const disclaimer = (
          <p key={ index } className={ this.namespace('footer-disclaimer') }>
            <span>
              { section.caption ? <FormattedMessage id={ section.caption } /> : section.title }{ ' ' }</span>
            <Link
              urlArgs={ props.urlArgs }
              href={ section.children[0].href }
              data-eid={ golfToEid(firstCaption, this.footerLinkEid) }>
              { firstChild.caption ? <FormattedMessage id={ firstCaption } /> : firstChild.title }
            </Link>
          </p>
        );
        sections.disclaimer = disclaimer;
        return disclaimer;
      } else if (~firstCaption.indexOf('LinkList')) {
        // legal links
        const links = (
          <ul key={ index } className={ this.namespace('list-inline', 'footer-legal-links') }>
            { this.buildLegalLinks(props, section.children, props.seechange ? this.namespace('list-inline-item') : void 0) }
          </ul>
        );
        sections.links = links;
        return links;
      } else if (sectionCaptionOrId && (~sectionCaptionOrId.indexOf('CopyrightMessage'))) {
        // copyright
        const copyright = (
          <p key={ index } className={ this.namespace('footer-copyright') }>
            { section.caption ? <FormattedHTMLMessage id={ section.caption } values={{ year: props.year }} /> : section.title }
          </p>
        );
        sections.copyright = copyright;
        return copyright;
      }
    }).filter(Boolean);

    if (props.seechange) {
      return this.buildSeeChangeLegalSection(props, sections);
    }
    return (
      <div className={ this.namespace('footer-legal') }>
        <div className={ this.namespace('container') }>
          { rows }
        </div>
      </div>
    );
  }

  buildSeeChangeLegalSection(props, sections) {
    return (
      <div className={ this.namespace('footer-seechange-legal') }>
        <div className={ this.namespace('container') }>
          <div className={ this.namespace('row') }>
            <div className={ this.namespace('col-lg-5', 'col-12', 'pull-lg-right') }>
              { sections.links }
            </div>
            <div className={ this.namespace('col-lg-7', 'col-12') }>
              { sections.copyright }
            </div>
            <div className={ this.namespace('col-lg-6', 'col-12', 'legal-right') }>
              { sections.disclaimer }
            </div>
          </div>
        </div>
      </div>
    );
  }

  /**
   * Builds the display hero for the Support Phone in the current market
   * @param   {Object} supportContacts Data structure normalized by @ux/manifest-render-service
   *                                  (i.e. ux.eldorado.data.props.supportContacts).
   * @returns {React.ReactNode} Fully realized Support Phone UX.
   */
  buildSupportPhone(supportContacts = {}) {
    const number = supportContacts.technicalSupportPhone || '';
    const phone = {
      href: 'tel:' + number.replace(/[()\s-]/g, ''),
      number
    };

    //
    // Disable max-len for this block because we need some things on the same
    // line to preserve spaces
    //
    /* eslint-disable max-len */
    if (this.isMobile) {
      return <div className={ this.namespace('support-short') }>
        <FormattedMessage id={ 'Sales:Footer:SupportBar:MobileCallSupportTeam' } /><a href={ phone.href }
          className={ this.namespace('support-footer-number') }>{ phone.number }</a>
      </div>;
    }

    return <div>
      <FormattedMessage id={ 'Sales:Footer:SupportBar:NeedHelp' } /> <FormattedMessage
        id={ 'Sales:Footer:SupportBar:CallSupportTeam' } /> <a href={ phone.href }
        className={ this.namespace('support-footer-number') }>{ phone.number }</a>
    </div>;
    /* eslint-enable max-len */
  }

  /**
   * Performs all of the expensive React VDOM creation (i.e. React.createElement calls)
   * for Elements that are **immutable w.r.t. to state.**
   * @param   {Object} props SalesFooterComponent propsType
   * @returns {Object} Fully expanded state to spread into the current state.
   */
  nextStateFromProps(props = {}) {
    const nextState = {
      legalSection: this.buildLegalSection(props),
      socialMediaLinks: this.buildSocialMediaLinks(props),
      supportPhone: this.buildSupportPhone(props.supportContacts)
    };

    const marketInfo = props.marketData[props.market];
    const displayTaxesAndFees = marketInfo &&
                                props.vatInclusion &&
                                (props.vatInclusion.indexOf(props.market) > -1) &&
                                marketInfo.displayTaxesAndFees;

    if (displayTaxesAndFees) {
      nextState.vatDisclaimer = <div className={ this.namespace('row', 'footer-vat-disclaimer') }>
        <div>
          <FormattedMessage id={ vatInclusionCodes[displayTaxesAndFees] } />
        </div>
        <hr className={ this.namespace('main-hr') } />
      </div>;
    }

    return nextState;
  }

  /**
   * Add eventListeners for breakpoints.
   *
   * @api private
   */
  componentDidMount() {
    this.breakpoint(...breakpoints);
  }

  /**
   * Remove the eventListeners of breakpoints.
   *
   * @api private
   */
  componentWillUnmount() {
    this.breakup();
  }

  /**
   * Updates the state associated with this component for the `nextProps`.
   * Primarily triggers slightly expensive object-creation for all VDOM
   * that remains constant until a new set of props is provided.
   * @param  {Object} nextProps Full set of props for this component.
   */
  componentWillReceiveProps(nextProps) {
    const nextState = this.nextStateFromProps(nextProps);
    this.setState(nextState);
  }

  renderGoDaddyGuides() {
    const props = this.props;

    return (
      <Link { ...props.urls.help }
        eid='uxp.hyd.sales_footer_seechange.godaddy_guides.link.click'
        className={ this.namespace('footer-godaddy-guides', 'clearfix') }>
        <span className={ this.namespace('guides-left') }>
          <span className={ this.namespace('godaddy-guides-line', 'guides-line-1') }>
            <FormattedMessage id='Sales:Footer:GoDaddyGuides:WeTrulyLoveTakingYourCall' />
          </span>
          <span className={ this.namespace('godaddy-guides-line', 'guides-line-2') }>
            <FormattedMessage id='Sales:Footer:GoDaddyGuides:LogoAlt'>
              { altText => <img src={ props.urls.cdn.href + 'ux/sales-header/gd-guides.svg' }
                style={{ height: '30px' }}
                alt={ altText } /> }
            </FormattedMessage>
          </span>
        </span>
        <span className={ this.namespace('godaddy-guides-line', 'guides-line-3', 'guides-right', 'float-lg-left') }>
          <FormattedMessage id='Sales:Footer:GoDaddyGuides:LogoAlt'>
            { altText => <img src={ props.urls.cdn.href + 'ux/sales-header/care-arrow.svg' }
              alt={ altText } /> }
          </FormattedMessage>
        </span>
      </Link>
    );
  }

  renderSeeChangeTopSection() {
    const { messages, env, seechange } = this.props;

    // Desktop rendering
    if (~['desktop', 'large'].indexOf(this.state.breakpoint)) {
      return (
        <div className={ this.namespace('footer-seechange-top-section') }>
          <div className={ this.namespace('container') }>
            <div className={ this.namespace('row') }>
              <div className={ this.namespace('footer-seechange-top-left', 'col-md-8') }>
                <EmailForm messages={ messages } env={ env } seechange={ seechange }/>
              </div>
              <div className={ this.namespace('footer-seechange-top-right', 'col-md-4') }>
                { this.renderGoDaddyGuides() }
              </div>
            </div>
          </div>
        </div>
      );
    }

    // Tablet and below rendering (re-orders sections)
    return (
      <div className={ this.namespace('footer-seechange-top-section') }>
        <div className={ this.namespace('container') }>
          <div className={ this.namespace('row') }>
            <div className={ this.namespace('footer-seechange-top-right', 'col-xs-12') }>
              { this.renderGoDaddyGuides() }
            </div>
            <div className={ this.namespace('footer-seechange-top-left', 'col-xs-12') }>
              <EmailForm messages={ messages } env={ env } seechange={ seechange }/>
            </div>
          </div>
        </div>
      </div>
    );
  }

  renderFooterLowerSeechange(socialMediaLinks, marketSelectorProps, currencySelectorProps) {
    const props = this.props;
    return (
      <React.Fragment>
        <div className={ this.namespace('footer-lower') }>
          <div className={ this.namespace('footer-lower-container') } >
            <Logo
              market={ props.market }
              messages={ props.messages }
              logos={ props.logos }
              href={ props.urls.home.href }
              privateLabelId={ 1 /* This is GoDaddy only */ }
              showCountryNameOnLogo={ false }
              spanClass={ this.namespace('logo') }
              className={ this.namespace('logo-link') }
              isMobile={ false /* Always show full logo here */ }
              includeGDHead={ false }
              id='footer-lower-logo'
            />
            <div className={ this.namespace('hr', 'hidden-lg-up') }/>
            <div className={ this.namespace('selectors') }>
              <MarketSelector { ...marketSelectorProps } />
              <CurrencySelector { ...currencySelectorProps } />
            </div>
            { socialMediaLinks && socialMediaLinks.length > 0 ?
              <div className={ this.namespace('follow-us') }>
                <ul className={ this.namespace('list-inline', 'social-media-links') }>
                  { socialMediaLinks }
                </ul>
              </div> : null }
          </div>
        </div>
        <div className={ this.namespace('hr', 'hidden-md-down') }/>
      </React.Fragment>
    );
  }

  renderFooterLower(socialMediaLinks, marketSelectorProps, currencySelectorProps) {
    const props = this.props;
    return (
      <div className={ this.namespace('footer-lower') }>
        <div className={ this.namespace('container') }>
          <div className={ this.namespace('row') }>
            <div className={ this.namespace('col-sm-12') }>
              <hr className={ this.namespace('main-hr') } />
            </div>
          </div>
          <div className={ this.namespace('row') }>
            <div className={ this.namespace('col-md-6', 'col-xs-12', 'selectors') }>
              <div className={ this.namespace('flex-row',
                'justify-content-center',
                'justify-content-md-start') }>
                <MarketSelector { ...marketSelectorProps } />
                <CurrencySelector { ...currencySelectorProps } />
              </div>
            </div>
            <div className={ this.namespace('col-md-6', 'col-xs-12') }>
              { socialMediaLinks && socialMediaLinks.length > 0 ?
                <div className={ this.namespace('follow-us') }>
                  { !props.seechange &&
                    <span className={ this.namespace('font-primary-bold') }>
                      <FormattedMessage id={ 'Sales:Footer:SocialMedia:Heading' } />
                    </span> }
                  <ul className={ this.namespace('list-inline', 'social-media-links') }>
                    { socialMediaLinks }
                  </ul>
                </div> : null }
            </div>
          </div>
        </div>
      </div>
    );
  }
  /**
   * Renders footer
   *
   * @api private
   * @returns {React.ReactNode} namespaced footer
   */
  render() {
    const props = this.props;
    const { seechange } = props;
    const {
      legalSection,
      socialMediaLinks,
      supportPhone,
      vatDisclaimer
    } = this.state;

    const marketSelectorProps = assign({
      urlArgs: props.urlArgs,
      manifest: props.manifest,
      eidPrefix: this.eidPrefix,
      urls: props.urls,
      messages: props.messages,
      currentMarket: props.market,
      numColumns: this.state.breakpoint === 'tablet' ? 2 : 4,
      currentPage: props.currentPage,
      lockScrollOnOpen: true,
      marketData: props.marketData
    }, props.marketSelector);

    const currencySelectorProps = {
      messages: props.messages,
      eidPrefix: this.eidPrefix,
      market: props.market,
      currency: props.currency,
      urls: props.urls,
      privateLabelId: props.privateLabelId
    };

    // This is a placeholder div that is used for Tealium's privacy
    // policy content injection. Do not modify this without speaking
    // to the Tealium team (see: #tealium-support on Slack)
    const tealiumPrivacyPlaceholder = <div id='gtm_privacy'></div>;

    return (
      <IntlProvider locale={ props.market } messages={ props.messages }>
        <div className={ this.namespace(this.classNames('footer', {
          'footer-seechange': seechange
        })) }>
          { props.features.footerSupportBar !== false &&
          <div className={ this.namespace('footer-support') }>
            <div className={ this.namespace('container') }>
              <h4>{ supportPhone }</h4>
            </div>
          </div> }
          { seechange && this.renderSeeChangeTopSection() }
          <div className={ this.namespace('footer-link-section') }>
            <div>
              { seechange &&
                <div className={ this.namespace('container') }>
                  { vatDisclaimer }
                  <div className={ this.namespace('row', 'footer-links') }>
                    { this.buildFooterLinks(props) }
                  </div>
                </div>
              }
              { !seechange && <div className={ this.namespace('container') }>
                { vatDisclaimer }
                <div className={ this.namespace('row', 'footer-links') }>
                  { this.buildFooterLinks(props) }
                </div>
                { props.features.footerEmailForm &&
                  <div className={ this.namespace('row', 'footer-apps') }>
                    <div className={ this.namespace('col-lg-5', 'col-sm-6') }>
                      <EmailForm messages={ props.messages } env={ props.env } seechange={ seechange }/>
                    </div>
                    <div className={ this.namespace('col-lg-7', 'col-lg-offset-3', 'col-sm-6', 'text-right') } />
                  </div> }
                { this.renderAppStoreSection() }
              </div> }
            </div>
            { seechange ?
              this.renderFooterLowerSeechange(socialMediaLinks, marketSelectorProps, currencySelectorProps) :
              this.renderFooterLower(socialMediaLinks, marketSelectorProps, currencySelectorProps)
            }
          </div>
          { legalSection }
          { this.renderGoDaddyStoreModal() }
          { tealiumPrivacyPlaceholder }
        </div>
      </IntlProvider>
    );
  }
}

SalesFooterComponent.propTypes = {
  ...NamespaceShape,
  messages: PropTypes.object.isRequired,
  market: PropTypes.string.isRequired,
  marketData: PropTypes.object,
  urls: PropTypes.object,
  year: PropTypes.number,
  currentPage: PropTypes.string,
  privateLabelId: PropTypes.number.isRequired,
  features: PropTypes.shape({
    footerEmailForm: PropTypes.bool,
    footerSupportBar: PropTypes.bool
  }),
  seechange: PropTypes.bool
};

SalesFooterComponent.defaultProps = {
  year: new Date().getFullYear(),
  marketData: {},
  privateLabelId: 1,
  features: {}
};

const NamespacedFooter = withNamespace(SalesFooterComponent);

export default class SalesFooter extends React.Component {
  render() {
    return (
      <NamespaceProvider namespace={ namespace() }>
        {
          // @ts-ignore
          <NamespacedFooter { ...this.props } />
        }
      </NamespaceProvider>
    );
  }
}
