// preact-portal/dist/preact-portal.js
(function(global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory(require('preact')) :
        typeof define === 'function' && define.amd ? define(['preact'], factory) :
        (global.preactPortal = factory(global.preact));
}(this, (function(preact) {
    'use strict';

    var classCallCheck = function(instance, Constructor) {
        if (!(instance instanceof Constructor)) {
            throw new TypeError("Cannot call a class as a function");
        }
    };

    var inherits = function(subClass, superClass) {
        if (typeof superClass !== "function" && superClass !== null) {
            throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
        }

        subClass.prototype = Object.create(superClass && superClass.prototype, {
            constructor: {
                value: subClass,
                enumerable: false,
                writable: true,
                configurable: true
            }
        });
        if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
    };

    var possibleConstructorReturn = function(self, call) {
        if (!self) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }

        return call && (typeof call === "object" || typeof call === "function") ? call : self;
    };

    var Portal = function(_Component) {
        inherits(Portal, _Component);

        function Portal() {
            classCallCheck(this, Portal);
            return possibleConstructorReturn(this, _Component.apply(this, arguments));
        }

        Portal.prototype.componentDidUpdate = function componentDidUpdate(props) {
            for (var i in props) {
                if (props[i] !== this.props[i]) {
                    return this.renderLayer();
                }
            }
        };

        Portal.prototype.componentDidMount = function componentDidMount() {
            this.renderLayer();
        };

        Portal.prototype.componentWillUnmount = function componentWillUnmount() {
            this.renderLayer(false);
            if (this.remote) this.remote.parentNode.removeChild(this.remote);
        };

        Portal.prototype.findNode = function findNode(node) {
            return typeof node === 'string' ? document.querySelector(node) : node;
        };

        Portal.prototype.renderLayer = function renderLayer() {
            var show = arguments.length <= 0 || arguments[0] === undefined ? true : arguments[0];

            if (this.props.into !== this.intoPointer) {
                this.intoPointer = this.props.into;
                if (this.into && this.remote) {
                    this.remote = preact.render(preact.h(PortalProxy, null), this.into, this.remote);
                }
                this.into = this.findNode(this.props.into);
            }

            this.remote = preact.render(preact.h(
                PortalProxy, {
                    context: this.context
                },
                show && this.props.children || null
            ), this.into, this.remote);
        };

        Portal.prototype.render = function render() {
            return null;
        };

        return Portal;
    }(preact.Component);

    var PortalProxy = function(_Component2) {
        inherits(PortalProxy, _Component2);

        function PortalProxy() {
            classCallCheck(this, PortalProxy);
            return possibleConstructorReturn(this, _Component2.apply(this, arguments));
        }

        PortalProxy.prototype.getChildContext = function getChildContext() {
            return this.props.context;
        };

        PortalProxy.prototype.render = function render(_ref) {
            var children = _ref.children;

            return children && children[0] || null;
        };

        return PortalProxy;
    }(preact.Component);

    return Portal;

})));



//////////////////
// WEBPACK FOOTER
// ../node_modules/preact-portal/dist/preact-portal.js
// module id = 45
// module chunks = 0


// WEBPACK FOOTER //
// node_modules/preact-portal/dist/preact-portal.js
