// @buzzfeed/buzzblocks/components/header-container/index.js
import { Application } from 't3js';
import solid from '../../js/services/solid';
import stickyRegistry from '../../js/services/sticky-registry';


/**
 * @overview
 * Wrapper module that monitors when the `bf_header_ui` header becomes sticky/non-sticky.
 * This is for sticky Awareness card/Spotlight and possible other sticky components
 * so they know where to position themselves on the page
 */
export default class HeaderContainer {
  constructor(context) {
    this.context = context;
    this.element = context.getElement();
    // @todo There seems to be a different sticky element on breakpoints below md?
    this.stickyContainer = this.element.querySelector('.js-sticky-container');
    this.stickyBreakpoint = 'md'; // as defined in `bf_header_ui` styles
  }

  init() {
    solid.breakpointObserver.on('match', this.onBreakpointChange);
    this.onBreakpointChange();
  }

  onBreakpointChange = () => {
    if (solid.isLargerThan(this.stickyBreakpoint)) {
      stickyRegistry.add(this.stickyContainer, {priority: 'high'});
    } else {
      stickyRegistry.remove(this.stickyContainer);
    }
  }

  destroy() {
    solid.breakpointObserver.off('match', this.onBreakpointChange);
    delete this.element;
    delete this.stickyContainer;
  }
}

Application.addModule('header-container', (context) => new HeaderContainer(context));
