// @forbes/fbs-ads/dist/criteo/criteo-bidding.service.js
import {
    loadScript
} from '@forbes/fbs-assets-loader';
// Criteo documentation: http://demo.criteo.com/support/publisher/cdb/#introduction
// See Criteo docs for more information on line item ranges
export var lineItemRanges = [{
        min: 0,
        max: 15,
        increment: '0.04'
    },
    {
        min: 15,
        max: 40,
        increment: '0.20'
    },
    {
        min: 40,
        max: 100,
        increment: '1.00'
    }
];
var CriteoBiddingService = (function() {
    function CriteoBiddingService() {}
    CriteoBiddingService.prototype.init = function() {
        return this.loadCriteo();
    };
    CriteoBiddingService.prototype.setTargeting = function() {
        this.criteo.SetDFPKeyValueTargeting();
    };
    CriteoBiddingService.prototype.fetchBids = function(adUnits, bidRanges) {
        var _this = this;
        return new Promise(function(resolve, reject) {
            _this.criteo.events.push(function() {
                try {
                    _this.criteo.SetLineItemRanges(_this.stringifyLineItemRanges(bidRanges));
                    _this.criteo.RequestBids(adUnits, resolve);
                } catch (e) {
                    reject(e);
                }
            });
        }).catch(function(error) {
            console.error(error);
            return;
        });
    };
    /**
     * stringifyLineItemRanges
     * @returns string
     *
     * Criteo expects a string for line item ranges with the following format:
     * 'min..max:increment;min..max:increment;min..max:increment'
     */
    CriteoBiddingService.prototype.stringifyLineItemRanges = function(bidRanges) {
        return bidRanges
            .map(function(itemRange) {
                return itemRange.min + ".." + itemRange.max + ":" + itemRange.increment;
            })
            .join(';');
    };
    CriteoBiddingService.prototype.loadCriteo = function() {
        var _this = this;
        return new Promise(function(resolve, reject) {
            loadScript('https://static.criteo.net/js/ld/publishertag.js', function() {
                _this.criteo = window['Criteo'] = window['Criteo'] || {};
                _this.criteo.events = _this.criteo.events || [];
                resolve();
            });
        });
    };
    return CriteoBiddingService;
}());
export {
    CriteoBiddingService
};


//////////////////
// WEBPACK FOOTER
// ./node_modules/@forbes/fbs-ads/dist/criteo/criteo-bidding.service.js
// module id = 249
// module chunks = 16
