// @guardian/consent-management-platform/lib/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var isRunningOnProd = function () {
    if (typeof document === 'undefined') {
        return false;
    }
    var domain = document.domain || '';
    var shortDomain = domain
        .split('.')
        .slice(-2)
        .join('.');
    return shortDomain === 'theguardian.com';
};
var isProd = isRunningOnProd();
var cmpDomain = function () {
    return isProd
        ? 'https://manage.theguardian.com'
        : 'https://manage.code.dev-theguardian.com';
};
var cmpLogsUrl = function () {
    return isProd
        ? 'https://consent-logs.guardianapis.com/report'
        : 'https://consent-logs.code.dev-guardianapis.com/report';
};
var iabVendorListUrl = function () {
    return isProd
        ? 'https://www.theguardian.com/commercial/cmp/vendorlist.json'
        : 'https://code.dev-theguardian.com/commercial/cmp/vendorlist.json';
};
exports.CMP_DOMAIN = cmpDomain();
exports.CMP_URL = exports.CMP_DOMAIN + "/consent";
exports.CMP_LOGS_URL = cmpLogsUrl();
exports.COOKIE_MAX_AGE = 395; // 13 months
exports.GU_AD_CONSENT_COOKIE = 'GU_TK';
exports.GU_COOKIE_NAME = 'guconsent';
exports.GU_COOKIE_VERSION = 1;
exports.IAB_VENDOR_LIST_URL = iabVendorListUrl();
exports.IAB_COOKIE_NAME = 'euconsent';
exports.IAB_CMP_ID = 112;
exports.IAB_CMP_VERSION = 1;
exports.IAB_CONSENT_SCREEN = 0;
exports.IAB_CONSENT_LANGUAGE = 'en';
exports.CMP_READY_MSG = 'readyCmp';
exports.CMP_SAVED_MSG = 'savedCmp';
exports.CMP_CLOSE_MSG = 'closeCmp';
exports.GU_PURPOSE_LIST = {
    purposes: [
        {
            id: 0,
            name: 'Essential',
            eventId: 'essential',
            description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
            integrations: [
                {
                    name: 'Ophan',
                    policyUrl: 'https://www.theguardian.com/info/privacy',
                },
                {
                    name: 'Confiant',
                    policyUrl: 'https://www.confiant.com/privacy',
                },
            ],
            alwaysEnabled: true,
        },
        {
            id: 1,
            name: 'Functional',
            eventId: 'functional',
            description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque quis malesuada ante.',
            integrations: [
                {
                    name: 'Pinterest',
                    policyUrl: 'https://policy.pinterest.com/',
                },
            ],
        },
        {
            id: 2,
            name: 'Performance',
            eventId: 'performance',
            description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque quis malesuada ante. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos.',
            integrations: [
                {
                    name: 'Sentry',
                    policyUrl: 'https://sentry.io/privacy/',
                },
                {
                    name: 'Google Analytics',
                    policyUrl: 'https://policies.google.com/privacy?hl=en-US',
                },
            ],
        },
    ],
};
