// @forbes/fbs-ads/dist/googletag/googletag-proxy.js
var GoogleTagProxy = (function() {
    function GoogleTagProxy(googletag) {
        this.enabled = false;
        this.queue = [];
        this.googletag = googletag;
    }
    GoogleTagProxy.prototype.push = function(cb) {
        this.queue.push(cb);
        if (this.enabled) {
            this.flush();
        }
    };
    GoogleTagProxy.prototype.pushFirst = function(cb) {
        this.queue.unshift(cb);
        if (this.enabled) {
            this.flush();
        }
    };
    GoogleTagProxy.prototype.flush = function() {
        while (this.queue.length) {
            this.googletag.cmd.push(this.queue.shift());
        }
    };
    GoogleTagProxy.prototype.enable = function() {
        this.enabled = true;
        this.flush();
    };
    GoogleTagProxy.prototype.disable = function() {
        this.enabled = false;
    };
    return GoogleTagProxy;
}());
export {
    GoogleTagProxy
};


//////////////////
// WEBPACK FOOTER
// ./node_modules/@forbes/fbs-ads/dist/googletag/googletag-proxy.js
// module id = 253
// module chunks = 16
