// preact-i18n/dist/preact-i18n.js
function n(n) {
    return n && "object" == typeof n && "default" in n ? n.default : n
}

function t(n) {
    return null !== n && void 0 !== n
}

function r(n, t) {
    var o = e({}, n);
    for (var i in t) t.hasOwnProperty(i) && (o[i] = n[i] && t[i] && "object" == typeof n[i] && "object" == typeof t[i] ? r(n[i], t[i]) : n[i] || t[i]);
    return o
}

function e(n, t) {
    for (var r in t) t.hasOwnProperty(r) && (n[r] = t[r]);
    return n
}

function o(n) {
    if (n = n || {}, "string" == typeof n && (n = n.split(",")), "join" in n) {
        for (var t = {}, r = 0; r < n.length; r++) {
            var e = n[r].trim();
            e && (t[e.split(".").pop()] = e)
        }
        return t
    }
    return n
}

function i(n, t) {
    return v = t || h, n && n.replace(/\{\{([\w.-]+)\}\}/g, u)
}

function u(n, t) {
    for (var r = t.split("."), e = v, o = 0; o < r.length; o++)
        if (null == (e = e[r[o]])) return "";
    return "string" == typeof e && e.match(/\{\{/) && (e = i(e, v)), e
}

function c(n, r, e, o, u, c) {
    r && (n = r + "." + n);
    var a = e && f(e, n);
    return (u || 0 === u) && a && "object" == typeof a && (a = a.splice ? a[u] || a[0] : 0 === u && t(a.none) ? a.none : 1 === u && t(a.one || a.singular) ? a.one || a.singular : a.some || a.many || a.plural || a.other || a), a && i(a, o) || c || null
}

function a(n, t, r) {
    var i = {};
    t = t || {}, n = o(n);
    for (var u in n)
        if (n.hasOwnProperty(u) && n[u]) {
            var a = n[u];
            r || "string" != typeof a ? a.nodeName === g && (a = e({
                fallback: a.children && a.children[0]
            }, a.attributes), i[u] = c(a.id, t.scope, t.dictionary, a.fields, a.plural, a.fallback)) : i[u] = c(a, t.scope, t.dictionary)
        }
    return i
}
var l = require("preact"),
    f = n(require("dlv")),
    p = /[?&#]intl=show/,
    s = function(n) {
        function t() {
            n.apply(this, arguments)
        }
        return n && (t.__proto__ = n), t.prototype = Object.create(n && n.prototype), t.prototype.constructor = t, t.prototype.getChildContext = function() {
            var n = this.props,
                t = n.scope,
                o = n.definition,
                i = n.mark,
                u = e({}, this.context.intl || {});
            return t && (u.scope = t), o && (u.dictionary = r(u.dictionary || {}, o)), (i || "undefined" != typeof location && String(location).match(p)) && (u.mark = !0), {
                intl: u
            }
        }, t.prototype.render = function(n) {
            var t = n.children;
            return t && t[0] || null
        }, t
    }(l.Component),
    d = function() {
        for (var n = [], t = arguments.length; t--;) n[t] = arguments[t];
        if (n.length > 1) {
            return y(n[0], n[1])
        }
        var r = n[0];
        return function(n) {
            return y(n, r)
        }
    },
    y = function(n, t) {
        return function(r) {
            return l.h(s, t || {}, l.h(n, r))
        }
    },
    h = {},
    v, g = function(n, t) {
        var r = n.id,
            e = n.children,
            o = n.plural,
            i = n.fields,
            u = t.intl,
            a = e && e[0],
            p = c(r, u && u.scope, u && u.dictionary, i, o, a);
        if (u && u.mark) {
            var s = "dictionary" + (u && u.scope ? "." + u.scope : "") + "." + r;
            return l.h("mark", {
                style: "background: " + (p ? f(u, s) ? "rgba(119,231,117,.5)" : "rgba(229,226,41,.5)" : "rgba(228,147,51,.5)"),
                title: r
            }, p)
        }
        return p
    },
    b = function(n, t) {
        var r = n.children,
            e = t.intl,
            o = r && r[0];
        return o && l.cloneElement(o, a(o.attributes, e, !0))
    },
    m = function(n) {
        return function(t) {
            return function(r, e) {
                var o = "function" == typeof n ? n(r) : n,
                    i = a(o, e.intl);
                return l.h(t, assign({}, r, i))
            }
        }
    };
d.intl = d, d.IntlProvider = s, d.Text = g, d.Localizer = b, d.withText = m, module.exports = d;


//////////////////
// WEBPACK FOOTER
// ../node_modules/preact-i18n/dist/preact-i18n.js
// module id = 5
// module chunks = 0


// WEBPACK FOOTER //
// node_modules/preact-i18n/dist/preact-i18n.js
