// @buzzfeed/bf-analytics/lib/common/binder/data_binder.js
import AttributeParser from "./bf_attribute_parser.js";

"use strict";

/**
 * Data-binder main class. Represented as a facade in front of all the modules.
 * @constructor
 */
function DataBinder() {
  this.domBinders = [];
  this.socialBinders = [];
  this.scrollBinders = [];
  this.generalBinders = [];
  this.videoBinders = [];

  this.parser = new AttributeParser();
  var i;

  /**
   * Initialize all the parts of binder by invocation of their 'init' function.
   * @returns {*} Returns nothing.
   */
  this.init = function () {
    for (i = 0; i < this.domBinders.length; i++) {
      this.domBinders[i].init();
    }
    for (i = 0; i < this.socialBinders.length; i++) {
      this.socialBinders[i].init();
    }
    for (i = 0; i < this.scrollBinders.length; i++) {
      this.scrollBinders[i].init();
    }
    for (i = 0; i < this.generalBinders.length; i++) {
      this.generalBinders[i].init();
    }
    for (i = 0; i < this.videoBinders.length; i++) {
      this.videoBinders[i].init();
    }
  };

  this.trigger = function (options) {
    if (options.type === "updateTargets") {
      this.scrollBinders.forEach(function (binder) {
        if (typeof binder.updateTargets === "function") {
          binder.updateTargets();
        }
      });
    }
  };
}

/**
 * @type {DataBinder}
 */
export default DataBinder;
