// @ux/utility-bar/src/account-tray/account-tray.js
/* jscs:disable requireSpacesInsideJsxExpressionContainers */

import React from 'react';
import PropTypes from 'prop-types';
import {
    TrayMenu
} from '@ux/tray-menu';
import {
    golf,
    breakpoints
} from '../enums';
import {
    NamespaceConsumer,
    NamespaceShape,
    withNamespace
} from '@ux/namespace-component';
import {
    IntlProvider,
    FormattedMessage
} from 'react-intl';
import UserDetails from './user-details';
import SignIn from './sign-in';
import CreateMyAccount from './create-account';
import QuickLinks from './quick-links';

/**
 * Render and expose the AccountTray as a functional Component.
 *
 * @param {Object} props Properties
 * @returns {ReactElement} Namespaced AccountTray.
 * @api public
 */
class AccountTray extends NamespaceConsumer {
    /**
     * Add eventListeners for breakpoints.
     *
     * @api private
     */
    componentDidMount() {
        this.breakpoint(...breakpoints);
    }

    /**
     * Remove the eventListeners of breakpoints.
     *
     * @api private
     */
    componentWillUnmount() {
        this.breakup();
    }

    render() {
            const props = this.props;
            let name;

            if (!this.isMobile) {
                if (props.loggedIn) {
                    // props.displayName is the short name (e.g. "Jane").
                    name = props.displayName;
                } else {
                    name = < FormattedMessage id = {
                        golf.login
                    }
                    />;
                }
            }

            const trayName = ( <
                span className = {
                    this.namespace('title-text')
                } > {
                    name
                } <
                /span>
            );

            return ( <
                IntlProvider locale = {
                    props.market
                }
                messages = {
                    props.messages
                } >
                <
                TrayMenu type = 'base'
                className = {
                    this.namespace(
                        this.classNames('account-tray', {
                            'is-pro': props.loggedIn && props.isPro,
                            'growthnav2': props.growthnav2
                        }))
                }
                name = {
                    trayName
                }
                data - eid = 'uxp.hyd.shared.utility_bar.account_tray.click' >
                <
                div className = {
                    this.namespace(props.seechange ? 'container-fluid' : 'container')
                } >
                <
                div className = {
                    this.namespace('row')
                } >
                <
                div className = {
                    this.namespace('col-xl-7', 'push-xl-5', 'col-lg-6', 'push-lg-6', 'mb-5')
                } >

                <
                UserDetails { ...props
                }
                />

                {
                    !props.loggedIn && ( < div className = {
                            this.namespace('row')
                        } >
                        <
                        SignIn { ...props
                        }
                        /> <
                        CreateMyAccount { ...props
                        }
                        /> <
                        /div>) } <
                        /div> <
                        QuickLinks { ...props
                        }
                        /> <
                        /div> <
                        /div> <
                        /TrayMenu> <
                        /IntlProvider>
                    );
                }
            }

            /**
             * Default properties for account tray.
             *
             * @type {Object}
             * @api public
             */
            AccountTray.defaultProps = {
                urls: {
                    products: 'https://mya.godaddy.com',
                    renewals: 'https://account.godaddy.com/billing?filter=expires&subFilter=90',
                    account: 'https://account.godaddy.com',
                    pro: 'https://pro.godaddy.com',
                    manageDomains: 'https://dcc.godaddy.com',
                    manageHosting: 'https://gateway.godaddy.com',
                    sso: {
                        o365Login: 'https://sso.godaddy.com',
                        webmailLogin: 'https://email.godaddy.com',
                        login: 'https://sso.godaddy.com?realm=idp&app=mya&path=',
                        logout: 'https://sso.godaddy.com/logout?realm=idp',
                        createAccount: 'https://sso.godaddy.com/account/create?realm=idp&app=www&path='
                    }
                },
                market: 'en-US',
                loggedIn: false,
                isPro: false,
                privateLabelId: 1
            };

            /**
             * Require properties.
             *
             * @type {Object}
             * @api public
             */
            AccountTray.propTypes = {
                ...NamespaceShape,
                growthnav2: PropTypes.bool,
                urls: PropTypes.object.isRequired,
                loggedIn: PropTypes.bool,
                privateLabelId: PropTypes.number,
                name: PropTypes.string,
                displayName: PropTypes.string,
                shopperId: PropTypes.string,
                isPro: PropTypes.bool,
                market: PropTypes.string.isRequired,
                messages: PropTypes.object.isRequired,
                vip: PropTypes.object
            };

            export default withNamespace(AccountTray);
