// @ux/utility-bar/src/account-tray/create-account.js
/* jscs:disable requireSpacesInsideJsxExpressionContainers */

import React from 'react';
import PropTypes from 'prop-types';
import { TrayTitle } from '@ux/tray-menu';
import { golf } from '../enums';
import eidBuilder from '../eid';
import { NamespaceConsumer, NamespaceShape, withNamespace } from '@ux/namespace-component';
import { IntlProvider, FormattedMessage } from 'react-intl';
import Link from '@ux/link';

/**
 * Render CreateMyAccount section.
 *
 * @Constructor
 * @class CreateMyAccount
 * @api public
 */
class CreateMyAccount extends NamespaceConsumer {
  constructor(props) {
    super(...arguments);

    this.eids = {
      createAccount: eidBuilder({ key: golf.createAccount, manifest: props.manifest })
    };
  }
  render() {
    const props = this.props;
    const eids = this.eids;

    return (
      <IntlProvider locale={ props.market } messages={ props.messages }>
        <div className={ this.namespace('col-xl-5', 'pull-xl-7', 'col-lg-6', 'pull-lg-6', 'col-md-6', 'sign-up', 'divider') }>
          <TrayTitle>
            <FormattedMessage id={ golf.newCustomerHeading } />
          </TrayTitle>
          <p>
            <FormattedMessage
              id={ golf.newCustomerMessage }
              values={{ privateLabelName: props.privateLabelName }}
            /> <FormattedMessage id={ golf.createAccountMessage } />
          </p>
          <Link className={ this.namespace('btn', 'btn-sm', 'btn-default-purchase') }
            urlArgs={ props.urlArgs }
            { ...props.urls.sso.createAccount }
            data-eid={ eids.createAccount }>
            <FormattedMessage id={ golf.createAccount } />
          </Link>
        </div>
      </IntlProvider>
    );
  }
}

/**
 * Require properties.
 *
 * @type {Object}
 * @api public
 */
CreateMyAccount.propType = {
  ...NamespaceShape,
  urls: PropTypes.object.isRequired,
  market: PropTypes.string.isRequired,
  messages: PropTypes.object.isRequired
};

export default withNamespace(CreateMyAccount);
