// @buzzfeed/buzzblocks/js/services/client/index.js
const $html = document.documentElement;
const autoPlayGifs = false;

/**
 * Feature detection
 */
const supportsSVG = (function supportsSVG() {
  return !!document.createElementNS &&
    !!document.createElementNS('http://www.w3.org/2000/svg', 'svg').createSVGRect;
}());

const supportsLocalStorage = (function supportsLocalStorage() {
  try {
    localStorage.setItem('ls_test', true);
    localStorage.removeItem('ls_test');
    return true;
  } catch(e) {
    return false;
  }
}());

/**
 * We need to make certain feature detections available to the CSS
 */
if (!supportsSVG) {
  $html.classList.add('no-svg');
}

if (!!navigator.userAgent.match(/Pinterest/i) || !!document.referrer.match(/pinterest.com/i)) {
  $html.classList.add('pinterest');
}

export default {
  /**
   * Detect if device has touch
   * @return {boolean} - has touch
   */
  hasTouch: (function() {
    const hasTouch = 'ontouchstart' in window;
    if (hasTouch) {
      $html.classList.add('has-touch');
    }
    return hasTouch;
  }()),

  /**
   * @property {String} - return the client os
   */
  os: (function() {
    if (navigator.userAgent.match(/android/i)) {
      $html.classList.add('is-android');
      return 'android';
    }

    if (navigator.userAgent.match(/iphone|ipad|ipod/i)) {
      $html.classList.add('is-ios');
      return 'ios';
    }

    return '';
  }()),

  /**
   * @property {String} - return a normalized referrer
   */
  referrer: (function() {
    var _r = document.referrer;
    var _qs = window.location.search;

    var inApp = navigator.userAgent.match(/fban|twitter|pinterest/i);
    var inAppReferrer = inApp ? inApp[0] : '';

    if (_qs.match('referrer=pinterest') || _r.match('pinterest') || inAppReferrer === 'pinterest') {
      return 'pinterest';
    }

    if (_qs.match('referrer=twitter') || _r.match('t.co') || inAppReferrer === 'twitter') {
      return 'tweet';
    }

    if (_qs.match('referrer=facebook') || _r.match('m.facebook') || inAppReferrer === 'fban') {
      return 'facebook';
    }

    return '';
  }()),

  /**
   * @property {boolean} whether LS is supported and accessible
   */
  supportsLocalStorage,

  /**
   * @returns {boolean} whether we should use video
   */
  isVideoSupported() {
    if (autoPlayGifs) {
      return true;
    }
    return this.os === 'android';
  },

  /**
   * @returns {boolean} whether we should autoplay video
   */
  autoplaysVideo() {
    // We use iphone-inline-video to enable this on iOS
    return autoPlayGifs && this.os !== 'android';
  },

  /**
   * Returns viewport height
   * @returns {Number} - viewport height in pixels
   */
  getViewportHeight() {
    return window.innerHeight || $html.clientHeight;
  },

  /**
   * Returns viewport width
   * @returns {Number} - viewport width in pixels
   */
  getViewportWidth() {
    return window.innerWidth || $html.clientWidth;
  }
};
