// @buzzfeed/buzzblocks/components/masthead/masthead.js
import {
    Application
} from 't3js';

import componentApi from '../../js/services/component-api';
import bzfd from '../../js/services/bzfd';
import prepareTemplate from '../../js/services/nunjucks-env';
import template from './masthead-fe.html';
import util from '../../js/services/util/standalone';

class Masthead {
    constructor(ctx) {
        this.config = ctx.getConfig();
        this.element = ctx.getElement();
    }
    get options() {
        return {
            origin: bzfd.webRoot,
            ...this.config
        };
    }
    render(masthead) {
        if (!masthead) {
            return;
        }

        const displayName = this.config.displayName || util.capitalize(masthead.title);

        this.element.innerHTML = prepareTemplate(template).render({
            masthead,
            title_tag: this.config.titleTag || 'h4',
            display_name: displayName,
            config: this.config,
        });
        this.element.setAttribute('aria-label', `${displayName} Masthead`);
    }
    init() {
        componentApi.get('masthead', this.options)
            .then(this.render.bind(this))
            .catch(err => console.error(err));
    }
}

Application.addModule('masthead', ctx => new Masthead(ctx));

export default Masthead;
