// @buzzfeed/buzzblocks/js/behaviors/one-link-post/index.js
import {
    Application
} from 't3js';

/**
 * Behavior for view position monitoring.
 */
class OneLinkPost {
    constructor(context) {
        this.context = context;
        this.element = context.getElement();
        this.linkToPost = this.element.querySelector('.js-card__link');
    }

    onclick(event) {
        const hasTarget = event && event.target;
        const isNotLink = hasTarget && !event.target.closest('a');
        if (this.linkToPost && hasTarget && isNotLink) {
            try {
                // pass event to the link
                const newEvent = new MouseEvent(event.type, event);
                this.linkToPost.dispatchEvent(newEvent);
            } catch {
                // fallback for IE
                const openInTab = event.metaKey && !this.linkToPost.getAttribute('target');
                if (openInTab) {
                    this.linkToPost.setAttribute('target', '_blank');
                }
                this.linkToPost.click();
                if (openInTab) {
                    this.linkToPost.removeAttribute('target');
                }
            }
        }
    }
}

export const name = 'one-link-post';

Application.addBehavior(name, ctx => new OneLinkPost(ctx));

export default OneLinkPost;
