// @buzzfeed/bf-analytics/lib/common/binder/embed_video/vimeo_universal_embed_video.js
/**
 * https://developer.vimeo.com/player/js-api
 */

"use strict";

function VimeoUniversalEmbedVideoHandler(executable) {
  this.inited = false;
  var _this = this;

  if (window.addEventListener) {
    window.addEventListener("message", onMessageReceived, false);
  }
  else {
    window.attachEvent("onmessage", onMessageReceived, false);
  }

  // Handle messages received from the player
  function onMessageReceived(event) {
    // Handle messages from the vimeo player only
    if ((/^https?:\/\/player.vimeo.com/).test(event.origin)) {
      _this.processVideo(event);
    }
  }

  this.init = function () {
    this.inited = true;
  };

  this.processVideo = function (video) {
    var videoData = JSON.parse(video.data);
    if (videoData.event === "ready"){
      this.registerListeners(videoData.player_id);
    }
    if (this.inited){
      switch (videoData.event) {
        case "play":
          window[executable].apply(this, ["track/video/play", { data: _this.getVideoData(videoData), }, ]);
          break;
        case "pause":
          window[executable].apply(this, ["track/video/pause", { data: _this.getVideoData(videoData), }, ]);
          break;
        case "finish":
          window[executable].apply(this, ["track/video/finish", { data: _this.getVideoData(videoData), }, ]);
          break;
      }
    }
  };

  this.registerListeners = function (playerId) {
    var player = document.getElementById(playerId);
    player.contentWindow.postMessage(JSON.stringify({ method: "addEventListener", value: "play", }), "*");
    player.contentWindow.postMessage(JSON.stringify({ method: "addEventListener", value: "pause", }), "*");
    player.contentWindow.postMessage(JSON.stringify({ method: "addEventListener", value: "finish", }), "*");
  };

  this.getVideoData = function(videoData){
    var iFrame = document.getElementById(videoData.player_id);
    return {
      id: videoData.player_id,
      duration: videoData.data.duration,
      currentTime: videoData.data.seconds,
      // Vimeo mobile embed player don't have "data-published" attribute on iframe element. It's on parent.
      published: iFrame.getAttribute("data-published") || iFrame.parentElement.getAttribute("data-published"),
      url: iFrame.getAttribute("src"),
      source: "vimeo",
    };
  };
}

export default VimeoUniversalEmbedVideoHandler;
