// @buzzfeed/bf-analytics/lib/common/binder/embed_video/youtube_iframe_embed_video.js
/**
 * https://developers.google.com/youtube/iframe_api_reference
 */

"use strict";

function YoutubeIframeEmbedVideoHandler(executable) {
  this.inited = false;
  var _this = this;

  if (!window.bfaYoutubeIframePlayerTracking) {
    window.bfaYoutubeIframePlayerTracking = function (player) {
      _this.processVideo(player);
    }
  }

  this.init = function () {
    this.inited = true;
  };

  this.processVideo = function (player) {
    var iFrame = player.getIframe();
    var playerId = iFrame.getAttribute("id");
    var published = iFrame.getAttribute("data-published");
    var url = iFrame.getAttribute("src");

    player.addEventListener("onStateChange", function (event) {
      if (!_this.inited){
        return;
      }

      var data = {
        id: playerId,
        duration: event.target.getDuration(),
        currentTime: event.target.getCurrentTime(),
        published: published,
        url: event.target.getVideoUrl(),
        source: "youtube",
        type: "embed",
      };

      // `event.data` is a current player's state.
      switch (event.data) {
        case 1:
          window[executable].apply(this, ["track/video/play", {data: data, }, ]);
          break;
        case 2:
          window[executable].apply(this, ["track/video/pause", {data: data, }, ]);
          break;
        case 0:
          window[executable].apply(this, ["track/video/finish", {data: data, }, ]);
          break;
        case 3:
          window[executable].apply(this, ["track/video/startedBuffering", {data: data, }, ]);
          break;
      }
    });

    player.addEventListener("onError", function (event) {
      if (!_this.inited){
        return;
      }

      var data = {
        id: playerId,
        published: published,
        url: url,
        source: "youtube",
        errorCode: event.data,
      };
      window[executable].apply(this, ["track/video/error", {data: data, }, ]);
    });
  }
}

export default YoutubeIframeEmbedVideoHandler;
