// @exp/exp-utils/schema/schemaHandler.js
'use strict';

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var object = require('../helper/object');

var SchemaHandler = function () {
  function SchemaHandler(schemaHelper) {
    _classCallCheck(this, SchemaHandler);

    this.schemaHelper = schemaHelper;
  }

  // Abstract


  SchemaHandler.prototype.preProcess = function preProcess() {};

  SchemaHandler.prototype.load = function load(schemaType, sinks, data, extras) {
    this.schemaType = schemaType;
    this.sinks = ['ALL'].concat(sinks);
    this.data = data;
    this.extras = extras ? extras : {};
    this.preProcess();
    this.process();
  };

  SchemaHandler.prototype.process = function process(sinkFnMap) {
    var _this = this;

    if (this.sinks && this.sinks.length > 0) {
      object.map(sinkFnMap, function (sink, fn) {
        if (_this.sinks.indexOf(sink) !== -1) {
          fn(_this.data[sink]);
        }
      });
    }
  };

  return SchemaHandler;
}();

module.exports = SchemaHandler;
