// @forbes/fbs-ads/dist/medianet/medianet.service.js
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
import { loadScript } from '@forbes/fbs-assets-loader';
/**
 * Handles Medianet Header Bidding
 */
var MedianetBiddingService = (function () {
    function MedianetBiddingService() {
        this.advBidxc = {
            setTargetingForAdUnitsGPTAsync: function () { },
            next: function () { }
        };
        this.channel = '';
        this.section = '';
    }
    Object.defineProperty(MedianetBiddingService.prototype, "isLoaded", {
        get: function () {
            return this.advBidxc.isLoaded;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Initialize Medianet Header Bidding
     */
    MedianetBiddingService.prototype.init = function (config) {
        var _a = config.params, _b = _a === void 0 ? {} : _a, _c = _b.channel, channel = _c === void 0 ? [] : _c, _d = _b.section, section = _d === void 0 ? [] : _d;
        this.setChannelSection({
            channel: channel,
            section: section
        });
        this.loadMedianet(config.mobile);
    };
    /**
     * Apply Targeting for Medianet Header Bidding
     */
    MedianetBiddingService.prototype.applyConfig = function (data) {
        if (!data) {
            return;
        }
        this.data = data || this.data;
        this.advBidxc.channel = data.channel || this.channel;
        this.advBidxc.section = data.section || this.section;
        this.advBidxc.version = 4.1;
        this.advBidxc.domain = window.location.hostname;
        window['advBidxc'] = __assign({}, window['advBidxc'], this.advBidxc);
    };
    /**
     * Prefetch bids
     */
    MedianetBiddingService.prototype.prefetchBids = function () {
        if (this.advBidxc) {
            this.advBidxc.next();
        }
    };
    /**
     * Load medianet script
     */
    MedianetBiddingService.prototype.loadMedianet = function (mobile) {
        var _this = this;
        // @TODO - This is some pretty haphazard code. Fix it later
        // If the application loads the medianet script earlier than ads bootstrap, tell us not to load it again
        // App handles running the callback at the appropriate time
        if ((window['forbes'] || {}).fastFetchMedianet) {
            window['forbes'].fastFetchMedianetCb(function () {
                _this.advBidxc = __assign({}, window['advBidxc'], _this.advBidxc);
                _this.applyConfig(_this.advBidxc);
            });
        }
        else {
            var cid = mobile ? '8CUQ517Y2' : '8CUX956JU';
            loadScript("https://contextual.media.net/bidexchange.js?cid=" + cid + "&https=1", function () {
                _this.advBidxc = __assign({}, window['advBidxc'], _this.advBidxc);
                _this.applyConfig(_this.advBidxc);
            });
        }
    };
    MedianetBiddingService.prototype.setChannelSection = function (channelSection) {
        var channel = channelSection.channel, section = channelSection.section;
        this.channel = (channel || []).join(',');
        this.section = (section || []).map(function (sectionName) { return sectionName.split(':')[1]; }).join(',');
        if (this.advBidxc) {
            this.advBidxc.channel = this.channel;
            this.advBidxc.section = this.section;
        }
        this.applyConfig({ channel: this.channel, section: this.section });
    };
    return MedianetBiddingService;
}());
export { MedianetBiddingService };


//////////////////
// WEBPACK FOOTER
// ./node_modules/@forbes/fbs-ads/dist/medianet/medianet.service.js
// module id = 254
// module chunks = 16
