// @forbes/fbs-ads/dist/googletag/googletag.interfaces.js
var Pubads = (function () {
    function Pubads() {
    }
    Pubads.prototype.collapseEmptyDivs = function () {
        return false;
    };
    Pubads.prototype.enableAsyncRendering = function () {
        return false;
    };
    Pubads.prototype.enableSingleRequest = function () { };
    Pubads.prototype.disableInitialLoad = function () { };
    Pubads.prototype.clearTargeting = function (id) { };
    Pubads.prototype.refresh = function (slots, options) { };
    Pubads.prototype.addEventListener = function (name, callback) { };
    Pubads.prototype.setSafeFrameConfig = function (config) { };
    Pubads.prototype.setTargeting = function (key, value) {
        return this;
    };
    Pubads.prototype.getTargeting = function (key) {
        return [];
    };
    Pubads.prototype.getTargetingKeys = function () {
        return [];
    };
    Pubads.prototype.setRequestNonPersonalizedAds = function (val) { };
    return Pubads;
}());
export { Pubads };
// tslint:disable-next-line:max-classes-per-file
var GoogleTag = (function () {
    function GoogleTag() {
        this.apiReady = false;
        this.pubadsReady = false;
        this.cmd = [];
        this.dummyAd = {
            setTargeting: function (key, value) {
                return this;
            },
            getTargeting: function (key) {
                return [];
            },
            getTargetingKeys: function () {
                return [];
            },
            clearTargeting: function (key) {
                return this;
            },
            addService: function (service) {
                return this;
            },
            getSlotElementId: function () {
                return '';
            }
        };
    }
    GoogleTag.prototype.pubads = function () {
        return new Pubads();
    };
    GoogleTag.prototype.enableServices = function () { };
    GoogleTag.prototype.disableInitialLoad = function () { };
    GoogleTag.prototype.defineSlot = function (siteZone, sizes, id) {
        return this.dummyAd;
    };
    GoogleTag.prototype.display = function (id) { };
    GoogleTag.prototype.destroySlots = function (slotArray) {
        return false;
    };
    return GoogleTag;
}());
export { GoogleTag };


//////////////////
// WEBPACK FOOTER
// ./node_modules/@forbes/fbs-ads/dist/googletag/googletag.interfaces.js
// module id = 252
// module chunks = 16
