// @ux/notifications/src/list-item.js
import React from 'react';
import { NamespaceConsumer, NamespaceShape, withNamespace } from '@ux/namespace-component';
import shapes from './shapes';

class NotificationListItem extends NamespaceConsumer {
  constructor() {
    super(...arguments);
    this.renderNotificationLink = this.renderNotificationLink.bind(this);
    this.renderStaticNotification = this.renderStaticNotification.bind(this);
  }

  renderNotificationLink(notification, href, classes) {
    const { content, eids } = notification;
    return (
      <a target='_blank'
        rel='noopener noreferrer'
        href={ href }
        data-eid={ eids && eids.click }
        className={ classes }>
        <span>{ content }</span>
      </a>
    );
  }

  renderStaticNotification(notification, classes) {
    return (
      <span className={ this.classNames(
        classes,
        this.namespace('no-link'))
      }>{ notification.content }</span>
    );
  }

  render() {
    const { notification } = this.props;
    const { priority, subcategory } = notification;
    const type = (subcategory === 'FailedBilling') ? 'creditCardExpiration' : '';
    const priorityClassMap = {
      Medium: 'medium',
      High: 'high'
    };
    const priorityClass = priorityClassMap[priority] || '';
    const priorityLabelClassName = this.namespace('priority', priorityClass, type);
    const classes = this.namespace('notifications-list-item', 'font-base');
    const { link } = notification;

    return (
      <li className={ classes }>
        {
          link ? this.renderNotificationLink(notification, link, priorityLabelClassName)
            : this.renderStaticNotification(notification, priorityLabelClassName)
        }
      </li>
    );
  }
}

NotificationListItem.propTypes =  {
  ...NamespaceShape,
  notification: shapes.notification.isRequired
};

export default withNamespace(NotificationListItem);
