// @forbes/fbs-tracking/dist/src/fbs-tracking/google-analytics/google-analytics.service.js
import { loadScript } from '@forbes/fbs-assets-loader';
var gtmContainers = {
    prod: 'GTM-NMQJM4',
    staging: 'GTM-KRRS9R'
};
/**
 * This is a service for tracking to GoogleAnalytics through GTM
 */
var GoogleAnalyticsService = /** @class */ (function () {
    function GoogleAnalyticsService() {
        this.initialized = false;
        this.dataLayer = window['dataLayer'] = window['dataLayer'] || [];
        this.containerId = gtmContainers.staging; // This is the default container ID
    }
    GoogleAnalyticsService.convertData = function (data) {
        if (data.eventLabel) {
            data['event label'] = data.eventLabel;
        }
        if (data.eventCategory) {
            data['event category'] = data.eventCategory;
        }
        if (data.eventAction) {
            data['event action'] = data.eventAction;
        }
        if (data.eventValue) {
            data['event value'] = data.eventValue;
        }
        delete data.eventLabel;
        delete data.eventCategory;
        delete data.eventAction;
        delete data.eventValue;
    };
    /**
     * This fires a GA Call via GTM
     */
    GoogleAnalyticsService.prototype.track = function (data) {
        if (!data) {
            return;
        }
        GoogleAnalyticsService.convertData(data);
        this.dataLayerPush(data);
        if (!this.initialized) {
            this.init();
        }
    };
    /**
     * Set the container to match environment
     */
    GoogleAnalyticsService.prototype.setEnvironment = function (environment) {
        this.containerId = gtmContainers[environment];
    };
    /**
     * This puts necessary GA Variables on the page.
     */
    GoogleAnalyticsService.prototype.init = function () {
        this.initialized = true;
        this.dataLayerPush({
            'gtm.start': new Date().getTime(),
            event: 'gtm.js'
        });
        window['analyticsEvent']
            = window['analyticsSocial']
                = window['analyticsVPV']
                    = window['analyticsClearVPV']
                        = window['analyticsForm']
                            = function () { };
        loadScript('https://www.googletagmanager.com/gtm.js?id=' + this.containerId, function () {
            var event = new CustomEvent('fbs-gtm-init');
            window.dispatchEvent(event);
        });
    };
    /**
     * Just a wrapper around dataLayer.push
     */
    GoogleAnalyticsService.prototype.dataLayerPush = function (data) {
        this.dataLayer.push(data);
    };
    return GoogleAnalyticsService;
}());
export { GoogleAnalyticsService };


//////////////////
// WEBPACK FOOTER
// ./node_modules/@forbes/fbs-tracking/dist/src/fbs-tracking/google-analytics/google-analytics.service.js
// module id = 267
// module chunks = 16
