// @ux/password/index.js
import {
    withNamespace
} from '@ux/namespace-component';
import {
    FormElement
} from '@ux/form-element';
import PropTypes from 'prop-types';
import React from 'react'; // eslint-disable-line no-unused-vars

/**
 * Component for rending a password element which can switch the password on and
 * off.
 *
 * @displayName Password
 * @class Password
 * @public
 */
export class Password extends FormElement {
    constructor(...args) {
        super(...args);

        const component = this;

        //
        // Conditionally ignore properties,
        //
        if (component.ignoreProps) {
            component.ignoreProps.push('show', 'hide');
        }

        component.type = 'password'; // Default input type.
        component._show = this.props.show || 'show'; // Label for when we're in password mode.
        component._hide = this.props.hide || 'hide'; // Label for when password is visible.

        //
        // By the time the component is processed, we already received and processed
        // the props so we can just correctly set the button so it's rendered as
        // intended.
        //
        component.button = component.props.button || component.select('_show');

        /**
         * Switches the button state.
         *
         * @param {Event} e DOM Event
         * @private
         */
        component.action = function action(e) {
            e.preventDefault();

            const hidden = component.select('type') === 'password';

            component.setState({
                button: hidden ? component.select('_hide') : component.select('_show'),
                type: hidden ? 'text' : 'password'
            });
        };
    }

    render() {
        return super.render();
    }
}

/**
 * Allowed properties, show and hide are used as action.
 *
 * @type {Object}
 * @public
 */
Password.propTypes = {
    /** props from Element in FormElement */
    ...FormElement.propTypes,
    /** Label that is displayed when the user can show his typed password. */
    show: PropTypes.node,
    /** Label that is shown when user can hide his showed password. */
    hide: PropTypes.node
};

export default withNamespace(Password);



// WEBPACK FOOTER //
// ./node_modules/@ux/password/index.js
