// @buzzfeed/site-tracking/common/global-defaults.js
import solid from '@buzzfeed/buzzblocks/js/services/solid/standalone';

export const GLOBAL_DEFAULTS = {
    mode: () => solid.isXsmall() || solid.isSmall() ? 'mobile' : 'desktop',
    os: () => {
        var OperatingSystems = {
            LINUX: 'Linux',
            MAC_OS: 'MacOs',
            UNIX: 'Unix',
            UNKNOWN: 'Unknown',
            WINDOWS: 'Windows',
        };
        var HOST_OS = '';

        if (navigator.appVersion.indexOf('Win') !== -1) {
            HOST_OS = OperatingSystems.WINDOWS;
        } else if (navigator.appVersion.indexOf('Mac') !== -1) {
            HOST_OS = OperatingSystems.MAC_OS;
        } else if (navigator.appVersion.indexOf('X11') !== -1) {
            HOST_OS = OperatingSystems.UNIX;
        } else if (navigator.appVersion.indexOf('Linux') !== -1) {
            HOST_OS = OperatingSystems.LINUX;
        } else {
            HOST_OS = OperatingSystems.UNKNOWN;
        }

        return HOST_OS;
    },
    referrer: () => window.document.referrer,
    screenResolution: () => window.screen.width + ' x ' + window.screen.height,
    userAgent: () => navigator.userAgent.toLowerCase(),
};
