// @buzzfeed/bf-analytics/lib/plugins/testMode.js
import {
    defaultsDeep,
} from "lodash";

/**
 * Test mode plugin
 * @module  plugins/testMode
 * @requires plugins/generalSettings
 * @exports testMode
 */

"use strict";

/***
 * @classdesc
 * Test mode plugin. Checks specified cookie and updates general settings if test mode is active.
 * @memberof module:plugins/testMode
 * @param {BFACore} core - instance of bfa core
 * @param {Object} settings - plugin settings
 * @returns {Void} - void
 * @constructor
 */
function TestModePlugin(core, settings) {
    settings = settings || {};
    this.defaultSettings = {
        cookieName: "bfa_test",
    };
    this.settings = defaultsDeep(settings, this.defaultSettings);
    var cookies = core.utils.readCookie(this.settings.cookieName);
    this._isTestMode = cookies.length > 0 && cookies[0].value === "true";
    this.updateGeneralSettings = function(generalSettings) {
        if (generalSettings.mode !== core.defs.bfaMode.test) {
            core.push("general/settings/update", {
                mode: core.defs.bfaMode.test,
            });
        }
    }
    if (this._isTestMode) {
        this.updateGeneralSettings(core._generalSettings);
        core._onGeneralSettingsChange.add(this.updateGeneralSettings, this);
    }
    return {};
}

/**
 * defaultId property required for initialization
 * @type {string}
 */
TestModePlugin.defaultId = "testMode";
TestModePlugin.require = ["generalSettings", ];
export default TestModePlugin;
